src/array.jl

```diff
@@ -19,6 +19,10 @@ function reftype(sz::Int)
     end
 end
 
+fixstringtype(T::Type) = T <: SubString || T === AbstractString ? String : T
+fixstringtype(T::Union) = Union{fixstringtype(T.a), fixstringtype(T.b)}
+fixstringtype(::Type{Union{}}) = Union{}
```

Comment:
Yes - `Union{}` case is tricky here. Thank you for handling this. The PR looks good.